### DISCUSSION TITLE
PBR materials. Normal map has no effect. bevy 0.9

### DISCUSSION THREAD

**Discussion by kybarg** [bevyengine/bevy#6824] [Q&A] [answered]
When i applying **normal map** to my meterial I see no effect. Am I doing anything wrong?
Also how can I adjust effect of normal map?

```rust
    let cobbelstone_material = materials.add(StandardMaterial {
        reflectance: 0.0,
        metallic: 1.0,
        perceptual_roughness: 0.95,
        base_color_texture: Some(asset_server.load("textures/cobbelstone/Cobblestone01_1K_BaseColor.png")),
        normal_map_texture: Some(asset_server.load("textures/cobbelstone/Cobblestone01_1K_Normal.png")),
        metallic_roughness_texture: Some(asset_server.load("textures/cobbelstone/Cobblestone01_1K_Roughness.png")),
        occlusion_texture: Some(asset_server.load("textures/cobbelstone/Cobblestone01_1K_AO.png")),
        ..default()
    });
```
    
![image](https://user-images.githubusercontent.com/4588778/205342809-392518da-ae11-4428-88aa-d877d1f3b9d0.png)

**Comment by nicopap** [comment, marked as answer]
Bevy requires the mesh vertices to have tangents sets to use normal maps, make sure you generate them as described in the `StandardMaterial` doc (<https://docs.rs/bevy/latest/bevy/pbr/struct.StandardMaterial.html#structfield.normal_map_texture>)

**Reply by kybarg** [reply]
Thank you, it worked. Sorry, i'm new to **rust** and **bevy**. 
Is there a way I can adjust __min__ and __max__ for norml map effect?

**Reply by nicopap** [reply]
> Thank you, it worked. Sorry, i'm new to **rust** and **bevy**. Is there a way I can adjust **min** and **max** for norml map effect?

Hmm. I don't know. I'll ask around

**Reply by nicopap** [reply]
I'd suggest you try to modify the normal map in an image editor. I think I managed to reduce the strength of the normal map by "reducing contrast"

**Reply by nicopap** [reply]
So the steps to reduce the strength of a normal map is:

1. Open the normal map texture in an image editor
2. Create a new layer with uniform color #8080ff
3. adjust the transparency of the original image layer to the percent you think is best
4. Save the new image and use the new image.

**Reply by kybarg** [reply]
What If I want it to be dynamic? I think there should be a way to adjsut it on shader level, like other engines

**Reply by nicopap** [reply]
From <https://discord.com/channels/691052431525675048/1048329297078800434/1048801698279325776>

> do other engines have a way to modulate that through a different parameter? Should bevy provide this option

> Yeah I think Unity has a material parameter for it.
>
> I don’t have a strong opinion on it, but I think it should be an import setting on the asset rather than a parameter on the shader. That makes for less calculations in the shader, and fewer parameters on the shader.
>
> From my experience it’s uncommon to want to use the same normal map at different intensities.

Bevy wouldn't be able to support that until we have an asset pre-processor. Like Lukas, I also think it isn't useful to change this at runtime, but I agree it should be made a parameter maybe in a future potential editor. The bevy PBR shader already has a lot of parameters, and adding more can reduce performance, so making it an asset preprocessing option seems ideal.

**Reply by kybarg** [reply]
@nicopap good point, thank you. I've used Unty before and in general have a little experiecne with 3d game development. Sorry for newby quiestions😂
